### PR TITLE
3D hover text fixes for date axes

### PR DESCRIPTION
**Pull request by etpinard**
fixes https://github.com/plotly/plotly.js/issues/1407

### CODE REVIEW

File: test/jasmine/tests/gl_plot_interact_test.js [hunk 1]
```diff
@@ -100,14 +100,18 @@ describe('Test gl plot interactions', function() {
...
-                expect(node[0].innerHTML).toEqual('x: 140.72', 'x val on hover');
-                expect(node[1].innerHTML).toEqual('y: −96.97', 'y val on hover');
-                expect(node[2].innerHTML).toEqual('z: −96.97', 'z val on hover');
+                var tspan = d3.selectAll('g.hovertext').selectAll('tspan')[0];
+                expect(tspan[0].innerHTML).toEqual(xLabel, 'x val on hover');
+                expect(tspan[1].innerHTML).toEqual(yLabel, 'y val on hover');
+                expect(tspan[2].innerHTML).toEqual(zLabel, 'z val on hover');
+            }
+
+            it('should have', function(done) {
```
Comment (alexcjohnson) [inline test/jasmine/tests/gl_plot_interact_test.js]: This isn't new with this PR, but this labeling is a little weird - it looks like you're including the message ('x val on hover' etc) as part of the test description, but doing that doesn't make it read out the right way either in the code or when the test fails.

The way I understand it, these tests should be described as:
```js
describe('Noun', function() {
    it('makes a sentence when you include the "it"', function() {
        expect(true).toBe(false, 'more info if useful pinpoint the problem');
    });
});
```

Which, when it fails, will make a message that starts with a sentence:

```
Chrome 56.0.2924 (Mac OS X 10.12.3) Noun makes a sentence when you include the "it" FAILED
        Expected true to be false, more info if useful pinpoint the problem.
            <traceback>
```

So in this case I'd have expected it to be something like:

```js
describe('gl plot interactions', function() {
    it('makes the right hover text and point data', function() {
        // I'm not sure if the messages add much, I typically only include them
        // when they include some variables either to determine which loop entry
        // we failed on or some input to the calculation that led to the failure
        // but having them doesn't detract from the test usability so you're
        // welcome to keep them if you find them useful
        expect(ptData.x).toBe('140.72');
    });
});
```

File: src/plots/gl3d/scene.js [hunk 2]
```diff
@@ -58,10 +58,10 @@ function render(scene) {
     }
 
     function formatter(axisName, val) {
-        if(typeof val === 'string') return val;
-
         var axis = scene.fullSceneLayout[axisName];
-        return Axes.tickText(axis, axis.c2l(val), 'hover').text;
+        var calendar = trace[axis._id.charAt(0) + 'calendar'];
```
Comment (alexcjohnson) [inline src/plots/gl3d/scene.js]: don't we actually want to use the axis calendar for this? (which is what you get if you omit calendar as an arg to `d2l`)

File: src/plots/gl3d/scene.js [hunk 3]
```diff
@@ -58,10 +58,10 @@ function render(scene) {
...
+
+        return Axes.tickText(axis, axis.d2l(val, 0, calendar), 'hover').text;
```
Comment (alexcjohnson) [inline src/plots/gl3d/scene.js]: I don't see a mock with category data in 3D, though it does work:
```js
Plotly.newPlot(gd,[{x:[1,2,3],y:['a','b','c'],z:['2014-01-01','2016-01-01','2015-01-01'], type:'scatter3d'}])
```
I *think* this code is compatible with category data, but can you include a test for it?

Comment (etpinard) [inline src/plots/gl3d/scene.js]: This one here 

https://github.com/plotly/plotly.js/pull/1414/files#diff-c406dc5b6d20f7f95206516178b5be36R148

should do it.

Comment (alexcjohnson) [inline src/plots/gl3d/scene.js]: Oh perfect. My turn for new 👓

File: src/plots/gl3d/scene.js (hunk 2 above)
Comment (etpinard) [inline src/plots/gl3d/scene.js]: So, if I understand correctly: hover labels should show dates with the calendar of the axes (as opposed to the calendar of the trace)?

Comment (alexcjohnson) [inline src/plots/gl3d/scene.js]: Yes. That's what we do in 2D, and I think it makes sense, particularly when you're plotting different traces each with its own calendar, you want to be able to compare them within the system they were plotted on, irrespective of what the source data looks like.

File: test/jasmine/tests/gl_plot_interact_test.js (hunk 1 above)
Comment (etpinard) [inline test/jasmine/tests/gl_plot_interact_test.js]: https://github.com/plotly/plotly.js/pull/1414/commits/e9e0c833818d43299a761472ed80ebbc0ebe55c7 - better?

File: src/plots/gl3d/scene.js (hunk 2 above)
Comment (etpinard) [inline src/plots/gl3d/scene.js]: https://github.com/plotly/plotly.js/pull/1414/commits/cbf1f8627451308042b1ac845c2a1b2a004f023a - looks ok?

Comment (alexcjohnson) [inline src/plots/gl3d/scene.js]: Perfect - thanks for the test.

**Comment by alexcjohnson** [comment]
💃 💃

**Pull request merged**